src/hotspot/share/services/virtualMemoryTracker.cpp

```diff
@@ -523,29 +525,58 @@ bool VirtualMemoryTracker::remove_released_region(address addr, size_t size) {
     }
   }
 
-  VirtualMemorySummary::record_released_memory(size, reserved_rgn->flag());
-
-  assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
-  if (reserved_rgn->base() == addr ||
-      reserved_rgn->end() == addr + size) {
+  if (size <= reserved_rgn->size() && reserved_rgn->contain_region(addr, size)) {
+    VirtualMemorySummary::record_released_memory(size, reserved_rgn->flag());
+    assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
+    if (reserved_rgn->base() == addr || reserved_rgn->end() == addr + size) {
       reserved_rgn->exclude_region(addr, size);
-    return true;
+      return true;
+    } else {
+      address top = reserved_rgn->end();
+      address high_base = addr + size;
+      ReservedMemoryRegion high_rgn(high_base, top - high_base,
+                                    *reserved_rgn->call_stack(), reserved_rgn->flag());
+
+      // Use original region for lower region
+      reserved_rgn->exclude_region(addr, top - addr);
+      LinkedListNode<ReservedMemoryRegion>* new_rgn = _reserved_regions->add(high_rgn);
+      if (new_rgn == nullptr) {
+        return false;
+      } else {
+        reserved_rgn->move_committed_regions(addr, *new_rgn->data());
+        return true;
+      }
+    }
   } else {
-    address top = reserved_rgn->end();
-    address high_base = addr + size;
-    ReservedMemoryRegion high_rgn(high_base, top - high_base,
-      *reserved_rgn->call_stack(), reserved_rgn->flag());
+    address end = addr+size;
+    size_t remaining = size;
+    LinkedListNode<ReservedMemoryRegion>* node_rgn = _reserved_regions->find_node(rgn);
+    // Segment the region into stripes to be further processed recursively
+    while (remaining > 0) {
+      ReservedMemoryRegion* remove_rgn = node_rgn->data();
+      assert(remove_rgn != nullptr, "NULL region");
+
+      node_rgn = node_rgn->next();
+      assert(remove_rgn->base() < node_rgn->data()->base(), "not ascending bases");
+
+      // Allow for the segment to be partially released
+      long offset = remove_rgn->base() - addr;
```

Comment:
Just a fly-by (no time to look at this closer, sorry): long is incorrect for Windows (LLP64). ptrdiff_t is better but has a number of caveats. hotspot usually uses pointer_delta(), with a preceding test for order.

Reply:
Thank you! Done.